<a href="https://colab.research.google.com/github/oumeimaghnimi/Huggingface_hub_learning/blob/master/Copie_de_Big_Transfer_timm_version2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This notebook tries to instantiate the code sample of timm repository:
https://rwightman.github.io/pytorch-image-models/models/big-transfer/

In [ ]:
!pip install  datasets transformers[sentencepiece]

In [ ]:
!python -m pip install --upgrade pip; pip install git-lfs
#git lfs install
!apt-get install git-lfs

In [ ]:
!git config --global user.email "oumeimaghnimi29@gmail.com"
!git config --global user.name "oumeima"

In [ ]:
!huggingface-cli login
#!transformers-cli login


        _|    _|  _|    _|    _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|_|_|_|    _|_|      _|_|_|  _|_|_|_|
        _|    _|  _|    _|  _|        _|          _|    _|_|    _|  _|            _|        _|    _|  _|        _|
        _|_|_|_|  _|    _|  _|  _|_|  _|  _|_|    _|    _|  _|  _|  _|  _|_|      _|_|_|    _|_|_|_|  _|        _|_|_|
        _|    _|  _|    _|  _|    _|  _|    _|    _|    _|    _|_|  _|    _|      _|        _|    _|  _|        _|
        _|    _|    _|_|      _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|        _|    _|    _|_|_|  _|_|_|_|

        
Username: oumeima
Password: 
Login successful
Your token: IdlHLMifuinaYGfKtmyBFlCKpztZaHklVWKqXdqRoEUskQkVjXntXYrAfMLLOusJwpovVMlBEDqTqnlEHwHlkdBbAZnyveUmAOAnIxSUOVQZqHiqjqRNMlsJJgyTmGBk 

Your token has been saved to /root/.huggingface/token


In [ ]:
!huggingface-cli repo create test_timm_succeed

In [ ]:
 !git clone https://huggingface.co/oumeima/test_timm_succeed

Cloning into 'test_timm_succeed'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 3 (delta 0), reused 0 (delta 0)
Unpacking objects: 100% (3/3), done.


In [ ]:
!cd test_timm_succeed && ls

'Big Transfer (BiT)_timm_test_to_huggingface.ipynb'


In [ ]:
!huggingface-cli  whoami  

oumeima


In [ ]:
!cd /content/test_timm_succeed; git add . ; git status;   git commit -m "First pretrained BiT version",
!cd /content/test_timm_succeed; git push  "https://IdlHLMifuinaYGfKtmyBFlCKpztZaHklVWKqXdqRoEUskQkVjXntXYrAfMLLOusJwpovVMlBEDqTqnlEHwHlkdBbAZnyveUmAOAnIxSUOVQZqHiqjqRNMlsJJgyTmGBk@https://huggingface.co/oumeima /test-timm-succeed.git"

#!cd /content/test_timm_succeed; set :repository,  "https://IdlHLMifuinaYGfKtmyBFlCKpztZaHklVWKqXdqRoEUskQkVjXntXYrAfMLLOusJwpovVMlBEDqTqnlEHwHlkdBbAZnyveUmAOAnIxSUOVQZqHiqjqRNMlsJJgyTmGBk@https://huggingface.co/oumeima /application.git"; git push
#!cd /content/test_timm_succeed; git push -f origin main  
#"https://IdlHLMifuinaYGfKtmyBFlCKpztZaHklVWKqXdqRoEUskQkVjXntXYrAfMLLOusJwpovVMlBEDqTqnlEHwHlkdBbAZnyveUmAOAnIxSUOVQZqHiqjqRNMlsJJgyTmGBk@https://huggingface.co/oumeima 

#git lfs status, 

 #set :repository,  "https://<OAuthToken>@github.com/username/application.git"
 #git push "https://$GITHUB_ACTOR:$GITHUB_TOKEN@github.com/$INPUT_REPO.git" gh-pages
#https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#github_token-secret:
 #Hey, I was able to get this working.  Here is an example: https://github.com/pkgjs/gh-pages/blob/master/entrypoint.sh Basically there are two environment vars you can use GITHUB_ACTOR and GITHUB_TOKEN.  Then use the https push url.  To get the token passed in you need to use the secrets as documen…

On branch main
Your branch is ahead of 'origin/main' by 1 commit.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
On branch main
Your branch is ahead of 'origin/main' by 1 commit.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
fatal: unable to access 'https://IdlHLMifuinaYGfKtmyBFlCKpztZaHklVWKqXdqRoEUskQkVjXntXYrAfMLLOusJwpovVMlBEDqTqnlEHwHlkdBbAZnyveUmAOAnIxSUOVQZqHiqjqRNMlsJJgyTmGBk@https://huggingface.co/oumeima /test-timm-succeed.git/': Could not resolve host: https


In [1]:
import os
os.getcwd()

'/content'

In [2]:
!pip install  timm

     |████████████████████████████████| 376 kB 5.2 MB/s 


In [3]:
import timm

In [4]:

#in timm. _init_ #from .models import create_model, list_models, is_model, list_modules, model_entrypoint, \
   # is_scriptable, is_exportable, set_scriptable, set_exportable, has_model_default_key, is_model_default_key, \
    # get_model_default_value, is_model_pretrained"'

 #timm.models._init_#   from .registry import register_model, model_entrypoint, list_models, is_model, list_modules, is_model_in_modules,\
    #has_model_default_key, is_model_default_key, get_model_default_value, is_model_pretrained


import timm
from pprint import pprint
model_names = timm.list_models(pretrained=True)
pprint(model_names)


['adv_inception_v3',
 'cait_m36_384',
 'cait_m48_448',
 'cait_s24_224',
 'cait_s24_384',
 'cait_s36_384',
 'cait_xs24_384',
 'cait_xxs24_224',
 'cait_xxs24_384',
 'cait_xxs36_224',
 'cait_xxs36_384',
 'coat_lite_mini',
 'coat_lite_small',
 'coat_lite_tiny',
 'coat_mini',
 'coat_tiny',
 'convit_base',
 'convit_small',
 'convit_tiny',
 'cspdarknet53',
 'cspresnet50',
 'cspresnext50',
 'deit_base_distilled_patch16_224',
 'deit_base_distilled_patch16_384',
 'deit_base_patch16_224',
 'deit_base_patch16_384',
 'deit_small_distilled_patch16_224',
 'deit_small_patch16_224',
 'deit_tiny_distilled_patch16_224',
 'deit_tiny_patch16_224',
 'densenet121',
 'densenet161',
 'densenet169',
 'densenet201',
 'densenetblur121d',
 'dla34',
 'dla46_c',
 'dla46x_c',
 'dla60',
 'dla60_res2net',
 'dla60_res2next',
 'dla60x',
 'dla60x_c',
 'dla102',
 'dla102x',
 'dla102x2',
 'dla169',
 'dm_nfnet_f0',
 'dm_nfnet_f1',
 'dm_nfnet_f2',
 'dm_nfnet_f3',
 'dm_nfnet_f4',
 'dm_nfnet_f5',
 'dm_nfnet_f6',
 'dpn68',
 'dpn

In [5]:
import timm
from pprint import pprint
model_names = timm.list_models('*resne*t*')
pprint(model_names)


['bat_resnext26ts',
 'cspresnet50',
 'cspresnet50d',
 'cspresnet50w',
 'cspresnext50',
 'cspresnext50_iabn',
 'eca_lambda_resnext26ts',
 'ecaresnet26t',
 'ecaresnet50d',
 'ecaresnet50d_pruned',
 'ecaresnet50t',
 'ecaresnet101d',
 'ecaresnet101d_pruned',
 'ecaresnet200d',
 'ecaresnet269d',
 'ecaresnetlight',
 'ecaresnext26t_32x4d',
 'ecaresnext50t_32x4d',
 'ens_adv_inception_resnet_v2',
 'gcresnet50t',
 'gcresnext26ts',
 'geresnet50t',
 'gluon_resnet18_v1b',
 'gluon_resnet34_v1b',
 'gluon_resnet50_v1b',
 'gluon_resnet50_v1c',
 'gluon_resnet50_v1d',
 'gluon_resnet50_v1s',
 'gluon_resnet101_v1b',
 'gluon_resnet101_v1c',
 'gluon_resnet101_v1d',
 'gluon_resnet101_v1s',
 'gluon_resnet152_v1b',
 'gluon_resnet152_v1c',
 'gluon_resnet152_v1d',
 'gluon_resnet152_v1s',
 'gluon_resnext50_32x4d',
 'gluon_resnext101_32x4d',
 'gluon_resnext101_64x4d',
 'gluon_seresnext50_32x4d',
 'gluon_seresnext101_32x4d',
 'gluon_seresnext101_64x4d',
 'ig_resnext101_32x8d',
 'ig_resnext101_32x16d',
 'ig_resnext101_

load a pretrained model:

In [6]:

model = timm.create_model('resnetv2_101x1_bitm', pretrained=True)
model.eval()


ResNetV2(
  (stem): Sequential(
    (conv): StdConv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
    (pad): ConstantPad2d(padding=(1, 1, 1, 1), value=0.0)
    (pool): MaxPool2d(kernel_size=3, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (stages): Sequential(
    (0): ResNetStage(
      (blocks): Sequential(
        (0): PreActBottleneck(
          (downsample): DownsampleConv(
            (conv): StdConv2d(64, 256, kernel_size=(1, 1), stride=(1, 1), bias=False)
            (norm): Identity()
          )
          (norm1): GroupNormAct(
            32, 64, eps=1e-05, affine=True
            (act): ReLU(inplace=True)
          )
          (conv1): StdConv2d(64, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
          (norm2): GroupNormAct(
            32, 64, eps=1e-05, affine=True
            (act): ReLU(inplace=True)
          )
          (conv2): StdConv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
          (norm3)

load and preprocess the image: 


In [7]:
import urllib
from PIL import Image
from timm.data import resolve_data_config
from timm.data.transforms_factory import create_transform

In [8]:
config = resolve_data_config({}, model=model)
transform = create_transform(**config)

url, filename = ("https://github.com/pytorch/hub/raw/master/images/dog.jpg", "dog.jpg")
urllib.request.urlretrieve(url, filename)
img = Image.open(filename).convert('RGB')
tensor = transform(img).unsqueeze(0) # transform and add batch dimension

/usr/local/lib/python3.7/dist-packages/torchvision/transforms/transforms.py:281: UserWarning: Argument interpolation should be of type InterpolationMode instead of int. Please, use InterpolationMode enum.
  "Argument interpolation should be of type InterpolationMode instead of int. "


/usr/local/lib/python3.7/dist-packages/torchvision/transforms/transforms.py:281: UserWarning: Argument interpolation should be of type InterpolationMode instead of int. Please, use InterpolationMode enum.
  "Argument interpolation should be of type InterpolationMode instead of int. "

  trial to correct the error:
  https://github.com/pytorch/vision/blob/master/torchvision/transforms/transforms.py:
  from .functional import InterpolationMode, _interpolation_modes_from_int

   
   line 281: /usr/local/lib/python3.7/dist-packages/torchvision/transforms/transforms.py:281:
    
                  or

   line 387: in https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional.py
   
   (it is not an error, just an information error):
     if isinstance(interpolation, int):
           warnings.warn(
               "Argument interpolation should be of type InterpolationMode      instead of int. "
            "Please, use InterpolationMode enum."
        )
        interpolation = _interpolation_modes_from_int(interpolation)



interpolation as an argument 


https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional.py


def _interpolation_modes_from_int(i: int) -> InterpolationMode:
    inverse_modes_mapping = {
        0: InterpolationMode.NEAREST,
        2: InterpolationMode.BILINEAR,
        3: InterpolationMode.BICUBIC,
        4: InterpolationMode.BOX,
        5: InterpolationMode.HAMMING,
        1: InterpolationMode.LANCZOS,
    }
    return inverse_modes_mapping[i]


pil_modes_mapping = {
    InterpolationMode.NEAREST: 0,
    InterpolationMode.BILINEAR: 2,
    InterpolationMode.BICUBIC: 3,
    InterpolationMode.BOX: 4,
    InterpolationMode.HAMMING: 5,
    InterpolationMode.LANCZOS: 1,
}


    






resolve_data_config:
    from timm.data import resolve_data_config
    
def resolve_data_config(args, default_cfg={}, model=None, use_test_size=False, verbose=False):
   


   timm/data/config.py: 
    # resolve interpolation method
    new_config['interpolation'] = 'bicubic'
    if 'interpolation' in args and args['interpolation']:
        new_config['interpolation'] = args['interpolation']
    elif 'interpolation' in default_cfg:
        new_config['interpolation'] = default_cfg['interpolation']   

get the model predictions:

In [9]:
import torch
with torch.no_grad():
    out = model(tensor)
probabilities = torch.nn.functional.softmax(out[0], dim=0)
print(probabilities)
print(probabilities.shape)
# prints: torch.Size([1000])

/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)


tensor([8.6920e-06, 1.0017e-05, 1.4372e-05, 5.0999e-05, 4.5306e-05, 3.9800e-05,
        9.8520e-06, 9.6706e-06, 3.3262e-05, 8.4749e-06, 8.8097e-06, 9.8926e-06,
        4.8570e-06, 3.7003e-06, 8.8064e-06, 1.2141e-05, 4.9568e-06, 1.6457e-05,
        6.4504e-06, 5.0538e-06, 1.0002e-05, 2.6593e-05, 7.1253e-06, 8.1279e-06,
        5.7542e-06, 7.9966e-06, 1.7420e-05, 3.9293e-06, 5.2244e-06, 2.7674e-05,
        7.6498e-06, 1.3327e-05, 1.1538e-05, 3.8164e-06, 3.7070e-06, 2.9407e-06,
        9.7026e-06, 2.7875e-06, 6.1393e-06, 6.1891e-06, 5.6028e-06, 6.1973e-06,
        5.2446e-06, 2.6525e-06, 5.7981e-06, 3.2827e-06, 1.4908e-05, 1.7386e-05,
        2.9888e-06, 4.2177e-06, 1.1562e-05, 4.2993e-06, 3.0088e-05, 6.3115e-06,
        9.6404e-06, 9.9693e-06, 7.5385e-06, 5.0012e-06, 1.6144e-05, 1.4105e-05,
        1.3371e-05, 4.4746e-06, 5.4517e-06, 1.0142e-05, 2.0313e-05, 9.8770e-06,
        1.3727e-05, 4.2800e-06, 8.9869e-06, 2.3762e-06, 1.5132e-05, 1.7925e-05,
        6.0713e-06, 1.6020e-05, 1.5472e-

In [10]:
# Get imagenet class mappings
import urllib
url, filename = ("https://raw.githubusercontent.com/pytorch/hub/master/imagenet_classes.txt", "imagenet_classes.txt")
urllib.request.urlretrieve(url, filename) 
with open("imagenet_classes.txt", "r") as f:
    categories = [s.strip() for s in f.readlines()]
    print(categories)
    print(len(categories))

['tench', 'goldfish', 'great white shark', 'tiger shark', 'hammerhead', 'electric ray', 'stingray', 'cock', 'hen', 'ostrich', 'brambling', 'goldfinch', 'house finch', 'junco', 'indigo bunting', 'robin', 'bulbul', 'jay', 'magpie', 'chickadee', 'water ouzel', 'kite', 'bald eagle', 'vulture', 'great grey owl', 'European fire salamander', 'common newt', 'eft', 'spotted salamander', 'axolotl', 'bullfrog', 'tree frog', 'tailed frog', 'loggerhead', 'leatherback turtle', 'mud turtle', 'terrapin', 'box turtle', 'banded gecko', 'common iguana', 'American chameleon', 'whiptail', 'agama', 'frilled lizard', 'alligator lizard', 'Gila monster', 'green lizard', 'African chameleon', 'Komodo dragon', 'African crocodile', 'American alligator', 'triceratops', 'thunder snake', 'ringneck snake', 'hognose snake', 'green snake', 'king snake', 'garter snake', 'water snake', 'vine snake', 'night snake', 'boa constrictor', 'rock python', 'Indian cobra', 'green mamba', 'sea snake', 'horned viper', 'diamondback', 

the top-5 predictions class names:

In [11]:

# Print top categories per image
top5_prob, top5_catid = torch.topk(probabilities, 5)
#print(top5_prob)
#print(top5_catid)
print(top5_prob.size(0))
print(len(top5_prob))
for i in range(top5_prob.size(0)):
    print(categories[top5_catid[i]], top5_prob[i].item())
    #print(categories[top5_catid[i]], top5_prob[i])  # avoid result  as:   Samoyed   tensor(0.9478)
# prints class names and probabilities like:
# [('Samoyed', 0.6425196528434753), ('Pomeranian', 0.04062102362513542), ('keeshond', 0.03186424449086189), ('white wolf', 0.01739676296710968), ('Eskimo dog', 0.011717947199940681)]

5
5
Samoyed 0.947767972946167
Pomeranian 0.008878438733518124
Arctic fox 0.006980232894420624
keeshond 0.0065886895172297955
white wolf 0.0050971307791769505


In [13]:
#All of the models in timm have consistent mechanisms for obtaining various types of features from the model for tasks besides classification.
#Penultimate Layer Features (Pre-Classifier Features)
#1_Unpooled
#method_1#forward_features()

import torch
import timm
m = timm.create_model('xception41', pretrained=True)
o = m(torch.randn(2, 3, 299, 299))
print(f'Original shape: {o.shape}')
o = m.forward_features(torch.randn(2, 3, 299, 299))
print(f'Unpooled shape: {o.shape}')

Original shape: torch.Size([2, 1000])
Unpooled shape: torch.Size([2, 2048, 10, 10])


In [14]:
#All of the models in timm have consistent mechanisms for obtaining various types of features from the model for tasks besides classification.
#Penultimate Layer Features (Pre-Classifier Features)
#1_Unpooled
#method_1#Create with no classifier and pooling
import torch
import timm
m = timm.create_model('resnet50', pretrained=True, num_classes=0, global_pool='')
o = m(torch.randn(2, 3, 224, 224))
print(f'Unpooled shape: {o.shape}')

Downloading: "https://github.com/rwightman/pytorch-image-models/releases/download/v0.1-weights/resnet50_ram-a26f946b.pth" to /root/.cache/torch/hub/checkpoints/resnet50_ram-a26f946b.pth


Unpooled shape: torch.Size([2, 2048, 7, 7])


In [15]:
#All of the models in timm have consistent mechanisms for obtaining various types of features from the model for tasks besides classification.
#Penultimate Layer Features (Pre-Classifier Features)
#1_Unpooled
#method_2# #Remove it later
import torch
import timm
m = timm.create_model('densenet121', pretrained=True)
o = m(torch.randn(2, 3, 224, 224))
print(f'Original shape: {o.shape}')
m.reset_classifier(0, '')
o = m(torch.randn(2, 3, 224, 224))
print(f'Unpooled shape: {o.shape}')

Downloading: "https://github.com/rwightman/pytorch-image-models/releases/download/v0.1-weights/densenet121_ra-50efcf5c.pth" to /root/.cache/torch/hub/checkpoints/densenet121_ra-50efcf5c.pth


Original shape: torch.Size([2, 1000])
Unpooled shape: torch.Size([2, 1024, 7, 7])


In [16]:
#All of the models in timm have consistent mechanisms for obtaining various types of features from the model for tasks besides classification.
#Penultimate Layer Features (Pre-Classifier Features)
#2_pooled##Create with no classifier
import torch
import timm
m = timm.create_model('resnet50', pretrained=True, num_classes=0)
o = m(torch.randn(2, 3, 224, 224))
print(f'Pooled shape: {o.shape}')


Pooled shape: torch.Size([2, 2048])


In [17]:
#All of the models in timm have consistent mechanisms for obtaining various types of features from the model for tasks besides classification.
#Penultimate Layer Features (Pre-Classifier Features)
#2_pooled##Remove it later

import torch
import timm
m = timm.create_model('ese_vovnet19b_dw', pretrained=True)
o = m(torch.randn(2, 3, 224, 224))
print(f'Original shape: {o.shape}')
m.reset_classifier(0)
o = m(torch.randn(2, 3, 224, 224))
print(f'Pooled shape: {o.shape}')

Downloading: "https://github.com/rwightman/pytorch-image-models/releases/download/v0.1-weights/ese_vovnet19b_dw-a8741004.pth" to /root/.cache/torch/hub/checkpoints/ese_vovnet19b_dw-a8741004.pth


Original shape: torch.Size([2, 1000])
Pooled shape: torch.Size([2, 1024])


In [18]:
#Multi-scale Feature Maps (Feature Pyramid)
#backbone network at multiple scales.
#A feature backbone can be created by adding the argument features_only=True to any create_model call. By default 5 strides will be output from most models (not all have that many), with the first starting at 2 (some start at 1 or 4).
#Create a feature map extraction model:
import torch
import timm
m = timm.create_model('resnest26d', features_only=True, pretrained=True)
o = m(torch.randn(2, 3, 224, 224))
for x in o:
  print(x.shape)

Downloading: "https://github.com/rwightman/pytorch-image-models/releases/download/v0.1-weights/gluon_resnest26-50eb607c.pth" to /root/.cache/torch/hub/checkpoints/gluon_resnest26-50eb607c.pth


torch.Size([2, 64, 112, 112])
torch.Size([2, 256, 56, 56])
torch.Size([2, 512, 28, 28])
torch.Size([2, 1024, 14, 14])
torch.Size([2, 2048, 7, 7])


In [19]:
#Multi-scale Feature Maps (Feature Pyramid)
#backbone network at multiple scales.
#A feature backbone can be created by adding the argument features_only=True to any create_model call. By default 5 strides will be output from most models (not all have that many), with the first starting at 2 (some start at 1 or 4).
#Query the feature information:
#After a feature backbone has been created, it can be queried to provide channel or resolution reduction information to the downstream heads without requiring static config or hardcoded constants.
# The .feature_info attribute is a class encapsulating the information about the feature extraction points:
import torch
import timm
m = timm.create_model('regnety_032', features_only=True, pretrained=True)
print(f'Feature channels: {m.feature_info.channels()}')
o = m(torch.randn(2, 3, 224, 224))
for x in o:
  print(x.shape)


Downloading: "https://github.com/rwightman/pytorch-image-models/releases/download/v0.1-weights/regnety_032_ra-7f2439f9.pth" to /root/.cache/torch/hub/checkpoints/regnety_032_ra-7f2439f9.pth


Feature channels: [32, 72, 216, 576, 1512]
torch.Size([2, 32, 112, 112])
torch.Size([2, 72, 56, 56])
torch.Size([2, 216, 28, 28])
torch.Size([2, 576, 14, 14])
torch.Size([2, 1512, 7, 7])


In [20]:
#Select specific feature levels or limit the stride
#There are to additional creation arguments impacting the output features.

#out_indices selects which indices to output
#output_stride limits the feature output stride of the network (also works in classification mode BTW)
#out_indices is supported by all models, but not all models have the same index to feature stride mapping. Look at the code or check feature_info to compare. The out indices generally correspond to the C(i+1)th feature level (a 2^(i+1) reduction). For most models, index 0 is the stride 2 features, and index 4 is stride 32.
#output_stride is achieved by converting layers to use dilated convolutions. Doing so is not always straightforward, some networks only support output_stride=32.

import torch
import timm
m = timm.create_model('ecaresnet101d', features_only=True, output_stride=8, out_indices=(2, 4), pretrained=True)
print(f'Feature channels: {m.feature_info.channels()}')
print(f'Feature reduction: {m.feature_info.reduction()}')
o = m(torch.randn(2, 3, 320, 320))
for x in o:
  print(x.shape)

Downloading: "https://imvl-automl-sh.oss-cn-shanghai.aliyuncs.com/darts/hyperml/hyperml/job_45402/outputs/ECAResNet101D_281c5844.pth" to /root/.cache/torch/hub/checkpoints/ECAResNet101D_281c5844.pth


Feature channels: [512, 2048]
Feature reduction: [8, 8]
torch.Size([2, 512, 40, 40])
torch.Size([2, 2048, 40, 40])


In [ ]:
#How do I finetune this model?
model = timm.create_model('resnetv2_101x1_bitm', pretrained=True, num_classes=NUM_FINETUNE_CLASSES)


#To finetune on your own dataset, you have to write a training loop or adapt timm's training script to use your dataset.
                                             #https://rwightman.github.io/pytorch-image-models/scripts/



In [ ]:
#How do I train this model?
    #You can follow the timm recipe scripts for training a new model afresh: https://rwightman.github.io/pytorch-image-models/scripts/

In [ ]:
from transformers import AutoModelForImageClassification
model =AutoModelForImageClassification.from_pretrained("")

model.save_pretrained("/content/test_timm_succeed")
model.push_to_hub("test-timm",use_auth_token="IdlHLMifuinaYGfKtmyBFlCKpztZaHklVWKqXdqRoEUskQkVjXntXYrAfMLLOusJwpovVMlBEDqTqnlEHwHlkdBbAZnyveUmAOAnIxSUOVQZqHiqjqRNMlsJJgyTmGBk")

You will need an authentication token with your Hugging Face credentials to use the `push_to_hub` method. Execute `huggingface-cli login` in your terminal or by uncommenting the following cell:

In [ ]:
import numpy as np

from datasets import load_dataset, load_metric
from transformers import (
    AutoModelForSequenceClassification,
    AutoTokenizer,
    DataCollatorWithPadding,
    Trainer,
    TrainingArguments,
)

In [ ]:
checkpoint = "bert-base-cased"

In [ ]:
raw_datasets = load_dataset("glue", "mrpc")

tokenizer = AutoTokenizer.from_pretrained(checkpoint)

def tokenize_function(examples):
    return tokenizer(examples["sentence1"], examples["sentence2"], truncation=True)

tokenized_datasets = raw_datasets.map(tokenize_function, batched=True)

model = AutoModelForSequenceClassification.from_pretrained(checkpoint, num_labels=2)

training_args = TrainingArguments(
    "finetuned-bert-mrpc",
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    learning_rate=2e-5,
    weight_decay=0.01,
    evaluation_strategy="epoch",
    logging_strategy="epoch",
    log_level="error",
    push_to_hub=True,
    push_to_hub_model_id="finetuned-bert-mrpc",
    # push_to_hub_organization="huggingface",
     push_to_hub_token="IdlHLMifuinaYGfKtmyBFlCKpztZaHklVWKqXdqRoEUskQkVjXntXYrAfMLLOusJwpovVMlBEDqTqnlEHwHlkdBbAZnyveUmAOAnIxSUOVQZqHiqjqRNMlsJJgyTmGBk",
)

data_collator = DataCollatorWithPadding(tokenizer)

metric = load_metric("glue", "mrpc")

def compute_metrics(eval_preds):
    logits, labels = eval_preds
    predictions = np.argmax(logits, axis=-1)
    return metric.compute(predictions=predictions, references=labels)

In [ ]:
trainer = Trainer(
    model,
    training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
    data_collator=data_collator,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics,
)

In [ ]:
trainer.train()

TrainOutput(global_step=690, training_loss=0.36729109390922215, metrics={'train_runtime': 180.7776, 'train_samples_per_second': 60.87, 'train_steps_per_second': 3.817, 'total_flos': 445479905606400.0, 'train_loss': 0.36729109390922215, 'epoch': 3.0})

## Push to hub from the Trainer directly

The `Trainer` has a new method to directly upload the model, tokenizer and model configuration in a repo on the [Hub](https://huggingface.co/). It will even auto-generate a model card draft using the hyperparameters and evaluation results!

In [ ]:
trainer.push_to_hub()

'https://huggingface.co/oumeima/finetuned-bert-mrpc/commit/63d0302ab1d8709b53bf0e216b486d95647ede77'

If you are using your own training loop, you can push the model and tokenizer separately (and you will have to write the model card yourself):

In [ ]:
# model.push_to_hub("finetuned-bert-mrpc")
# tokenizer.push_to_hub("finetuned-bert-mrpc")

## You can load your model from anywhere using from_pretrained!

In [ ]:
from transformers import AutoModelForSequenceClassification

model_name = "sgugger/finetuned-bert-mrpc"
model = AutoModelForSequenceClassification.from_pretrained(model_name)

## You can use your model in a pipeline!

In [ ]:
from transformers import pipeline

classifier = pipeline("text-classification", model=model_name)

In [ ]:
classifier("My name is Sylvain. [SEP] My name is Lysandre")

## Updating a problematic file is super easy!

In [ ]:
model.config.label2id = {"not equivalent": 0, "equivalent": 1}

In [ ]:
model.config.id2label = {0: "not equivalent", 1: "equivalent"}

In [ ]:
model.config.push_to_hub("finetuned-bert-mrpc")

In [ ]:
classifier = pipeline("text-classification", model=model_name)

classifier("My name is Sylvain. [SEP] My name is Lysandre")